apps/arweave/src/ar_footprint_record.erl

```diff
@@ -0,0 +1,200 @@
+    the last 24 chunks of the first entropy footprint and chunks 20-48 of the second
+    entropy footprint. These chunks are from the first partition. The offset of the chunks
+    from the second partition is shifted by the number of chunks in the replica 2.9
+    entropy generated per partition (which is slightly bigger than the number of chunks
+    that can fit in the 3.6 TB partition).
+
+    Note that Packing does not have to be replica_2_9. We maintain this record
+    for any packing so that it is convenient to serve the data to any client.
+""".
+
+%%%===================================================================
+%%% Public interface.
+%%%===================================================================
+
+%% @doc Add a chunk to the footprint record.
+-spec add(PaddedOffset :: non_neg_integer(), Packing :: term(), StoreID :: string()) -> ok.
+add(PaddedOffset, Packing, StoreID) ->
+	Offset = get_offset(PaddedOffset),
+	ar_sync_record:add(Offset, Offset - 1, Packing, ar_data_sync_footprints, StoreID).
+
+%% @doc Get the offset of a chunk in the footprint record.
+-spec get_offset(PaddedOffset :: non_neg_integer()) -> non_neg_integer().
+get_offset(PaddedOffset) ->
+	FootprintsPerPartition = get_footprints_per_partition(),
+	FootprintSize = get_footprint_size(),
+	Partition = ar_replica_2_9:get_entropy_partition(PaddedOffset),
+	PartitionStartOffset = Partition * FootprintsPerPartition * FootprintSize,
+	SliceIndex = ar_replica_2_9:get_slice_index(PaddedOffset),
+	Footprint = get_footprint(PaddedOffset),
+	PartitionStartOffset + FootprintSize * Footprint + SliceIndex + 1.
+
+%% @doc Get the chunk's footprint's number, >= 0, < the maximum number of footprints
+%% in a partition.
+-spec get_footprint(Offset :: non_neg_integer()) -> non_neg_integer().
+get_footprint(Offset) ->
+	EntropyIndex = ar_replica_2_9:get_entropy_index(Offset, 0),
+	EntropyIndex div ?COMPOSITE_PACKING_SUB_CHUNK_COUNT.
+
+%% @doc Get the footprint bucket number of a chunk.
+-spec get_footprint_bucket(Offset :: non_neg_integer()) -> non_neg_integer().
+get_footprint_bucket(Offset) ->
+	get_offset(Offset) div ?NETWORK_FOOTPRINT_BUCKET_SIZE.
+
+%% @doc Get the synced footprint intervals of a chunk.
+-spec get_intervals(
+	Partition :: non_neg_integer(),
+	Footprint :: non_neg_integer(),
+	StoreID :: string()
+) -> term().
+get_intervals(Partition, Footprint, StoreID) ->
+	get_intervals(Partition, Footprint, any, StoreID).
+
+%% @doc Get the synced footprint intervals of a chunk.
+-spec get_intervals(
+	Partition :: non_neg_integer(),
+	Footprint :: non_neg_integer(),
+	Packing :: term(),
+	StoreID :: string()
+) -> term().
+get_intervals(Partition, Footprint, Packing, StoreID) ->
+	FootprintSize = get_footprint_size(),
+	FootprintsPerPartition = ?REPLICA_2_9_ENTROPY_COUNT div ?COMPOSITE_PACKING_SUB_CHUNK_COUNT,
+	PartitionStartOffset = Partition * FootprintsPerPartition * FootprintSize,
```

Comment:
I think this calculation will cause the partition start offset to drift further and further from the recall partition start offset as we get to higher partitions? If I've read it correctly?

My understanding is that we now have the entropy start and partition start mostly aligned for all partitions (entropy start might start a partial chunk later so we don't split buckets, but no more than one chunk later i think)

Reply:
You mean, we could reduct the actual FootprintSize a bit, and the same logic will keep working?

Reply:
I was thinking that PartitionStartOffset should always be set to `Partition * ?PARTITION_SIZE` (i.e. 3.6TB * partition number)

But I may be misunderstanding the intent of `PartitionStartOffset`

Reply:
@JamesPiechota I've updated the logic such that the footprint record with all the data will not have holes between partitions (assuming there are no holes in the data, off course) by grounding it on the fact every partition has 3600000221184 div 262144 buckets.